### PR TITLE
Implement seatop_touch

### PR DESCRIPTION
**Pull request by stacyharper**
Atm we got issue with the touch position sent to the clients. While
holding contact, leaving the initial container will continue to send
motion event to the client but with the new local position from the new
container.

This seatop goal is to send the position of the touch event, relatively
to the initial container layout position.

This will solve: https://github.com/swaywm/sway/issues/6454

I'm pretty new to Wayland and even more with the Sway codebase.

### CODE REVIEW

File: sway/input/seatop_touch.c
Comment (Eluminae) [inline sway/input/seatop_touch.c:64]: This make sway to crash. I'm not sure yet how to release cleanly the seatop.

**Comment by Eluminae** [comment]
For record, I firstly tried to rely on seatop_down but it use cursor as much as possible. As touches can be multiple, I think we have to implement a new dedicated seatop.

File: sway/input/seatop_touch.c
Comment (kennylevinsen) [reply sway/input/seatop_touch.c:64]: To change back to the default seatop, use `seatop_begin_default`. See `handle_button` in `sway/input/seatop_down.c` for an example.

Comment (Eluminae) [reply sway/input/seatop_touch.c:64]: Yeah, way better ! thanks !

**Comment by Eluminae** [comment]
Oops, sorry I missclicked on ready for review, which is not !

**Comment by kennylevinsen** [comment]
Just click "Convert to draft" in upper right near the reviewer section.

File: sway/input/seatop_touch.c
Comment (Xyene) [inline sway/input/seatop_touch.c:37]: Missing check for allocation failure.

File: sway/input/cursor.c
Comment (Xyene) [inline sway/input/cursor.c:517]: I think this `wlr_seat_touch_notify_up` belongs in the seatop, now that we have one.

File: sway/input/seatop_touch.c
Comment (Xyene) [inline sway/input/seatop_touch.c:44]: We should handle sending `wlr_seat_touch_notify_down` here instead of in `cursor.c` now.

**Review by Xyene — CHANGES_REQUESTED** [review]
Thanks for working on this. I've left a first round of comments.

File: sway/input/seat.c
Comment (Eluminae) [reply sway/input/seat.c]: Oh nice ! Thats a point I was not understanding when looking at other seatop patches !

File: sway/input/seatop_touch.c
Comment (Eluminae) [reply sway/input/seatop_touch.c:38]: added

File: sway/input/cursor.c
Comment (Eluminae) [reply sway/input/cursor.c]: I moved those notification and focus action in the seatop

File: sway/input/seatop_touch.c
Comment (Eluminae) [reply sway/input/seatop_touch.c:45]: This one too

**Review by Xyene — COMMENTED** [review]
Just a bunch of small nits, thanks for working on this.

I think with the current implementation, the following things are true:

1. the cursor can be moved by pointer input, but while a finger is held down on the screen, the surface under the pointer will not receive movement events (see [this](https://github.com/swaywm/sway/blob/bb3fd0abc56f39e35dc7f4e86f25da1b4a6efbd7/sway/input/cursor.c#L378-L380))
2. pointer buttons likewise are no-ops while a finger is held down on the screen

2\) is fine, but I think we should resolve 1\). To do that, I think you'd want to merge the`seatop_down` and `seatop_touch_down` files, so that the touch down seatop can have pointer/tablet motion handlers. It's fine if `struct seatop_down_event` has a `struct wl_list point_events` that's only ever used for touch input.

**Comment by Eluminae** [comment]
I pushed fixes you mentioned.

> is fine, but I think we should resolve 1). To do that, I think you'd want to merge theseatop_down and seatop_touch_down files, so that the touch down seatop can have pointer/tablet motion handlers. It's fine if struct seatop_down_event has a struct wl_list point_events that's only ever used for touch input.

I agree that those pointer move still must be send to the client.

I tried you solution. The thing is that the seatop_down pointer_motion think we pressed the pointer to begin. It rely on the cursor initial position. I think it make no sense to handle it that way cause this really is a touch event.

I think we should instead implement a pointer_motion in the seatop_touch to just notify with the current cursor position. 

What do you think about this ?

**Comment by Eluminae** [comment]
Okay I thought a little more about this and in fact you right. I have to send cursor position relative to the cont position. I just have to setup the initial cursor position in my begin_touch_down correctly.

**Comment by Eluminae** [comment]
Okay I tried to merge both. I let this change in a dedicated commit.

I dont think this is a good idea and it introduce more complexity to this simple seatop_touch. 

Plus anyway. the cursor disappear when I use the touch screen. I have to click with the mouse to display it again. And this action stop the previous seatop. I think we should just drop this commit :S

File: sway/input/seatop_down.c
Comment (Xyene) [inline sway/input/seatop_down.c:197]: We can reduce this duplication by:

```c
seatop_begin_down(seat, con, sx, sy);
handle_touch_down(seat, event);
```

You can either get rid of the `time_msec` parameter to `seatop_down` (it's not needed anywhere), or just pass 0 to it for now.

That'll also raise floating containers on touch, which we want.

File: sway/input/seatop_down.c
Comment (Xyene) [inline sway/input/seatop_down.c:69]: Nit: please keep lines < 80 columns wide. (Also line 75.)

File: sway/input/seat.c
Comment (Xyene) [inline sway/input/seat.c:1556]: Nit: move after the pointer funcs.

**Review by Xyene — COMMENTED** [review]
Thanks, code mostly looks good.

Regarding mixed pointer/touch input, there isn't really a consensus between other operating systems.

On Windows, the first touch point also updates the cursor position. If the pointer is moved at the same time as the first touch point, the cursor jitters around. This isn't the behavior Sway implements.

On Chrome OS, touch input does not update the cursor position, and the pointer is free to move the cursor. The cursor is hidden on touch down, but becomes visible again if the pointer moves. This is the intended behavior in Sway (see code snippets below, this allows e.g. scrolling a webpage with one hand while also using the pointer to click links).

https://github.com/swaywm/sway/blob/daf9ad1af73a0358fdabc29ca5817511c16e15a4/sway/input/cursor.c#L398
https://github.com/swaywm/sway/blob/daf9ad1af73a0358fdabc29ca5817511c16e15a4/sway/input/cursor.c#L316-L319

If it doesn't work like that something is broken in the code above (it should not require you to click the pointer to make the cursor visible, just move it), but that's not necessary to tackle in this PR unless you feel like it.

**Comment by Eluminae** [comment]
Okay I've done everything you said. It looks to works at least for the touch part. 

The cursor sometime display itself when using the mouse, and sometime it stay hidden. I still dont know really why. Anyway, it does not send motion event to the client. There probably is a case we want to handle in the pointer_motion in cursor.c.

Clicking with the mouse break the seatop but that is probably what we expect. Starting the seatop with a mouse click, then add touches, then up touches will stop the seatop. It may be not what we expect.

**Comment by Xyene** [comment]
> The cursor sometime display itself when using the mouse, and sometime it stay hidden.

That's weird, but likely unrelated to this PR. Unfortunately, I don't own a touchscreen device with which to test this myself.

> Anyway, it does not send motion event to the client.

Hmm, are you sure? `pointer_motion` will call into `seatop_pointer_motion`,

https://github.com/swaywm/sway/blob/bb3fd0abc56f39e35dc7f4e86f25da1b4a6efbd7/sway/input/cursor.c#L378-L380

which should be sending motion events here

https://github.com/swaywm/sway/blob/bb3fd0abc56f39e35dc7f4e86f25da1b4a6efbd7/sway/input/seatop_down.c#L49

> Clicking with the mouse break the seatop but that is probably what we expect.

Maybe. I don't have a good sense for how that is to use (in Chrome OS, it doesn't, you can click multiple times with the pointer while scrolling via touch), but it doesn't sound terrible.

> Starting the seatop with a mouse click hold, then add touches, then up touches will stop the seatop. It may not be what we expect.

Yeah, that's a little weird. Not sure we can do a lot about this with the current structure. The seatop split works well for pointer/tablet input which is single-point, but doesn't generalize well to multi-touch input. Fixing this would be a lot of work.

File: sway/input/seatop_default.c
Comment (Xyene) [inline sway/input/seatop_default.c:643]: Nit: extra newline.

File: sway/input/seatop_down.c
Comment (Xyene) [inline sway/input/seatop_down.c:197]: Nit: extra newline.

**Review by Xyene — COMMENTED** [review]
Otherwise LGTM.

File: sway/input/seatop_default.c
Comment (Eluminae) [reply sway/input/seatop_default.c:643]: I really like those crazy guys...

**Comment by Eluminae** [comment]
At least now client can implement correctly touch holded event :]

**Review by Xyene — DISMISSED** [review]
Thanks! I'll wait for @emersion to merge, because he actually owns a touchscreen device 🙂

**Comment by Eluminae** [comment]
Okay nice ! It is very easy to test with wev. 

Thanks a lot for your reviews and assistance on this @Xyene !

**Comment by stacyharper** [comment]
Okay maybe it is not that ready yet. I got crashes while touching the swaybar.

**Comment by stacyharper** [comment]
`con` may be NULL. I have to check it before `container_raise_floating`

**Comment by Xyene** [comment]
This should be fixed in https://github.com/swaywm/sway/pull/6418 (there are some other cases that could crash too, see that PR for details). We should get that merged before this one.

**Comment by stacyharper** [comment]
Oh yes, that's exactly that. I'll have both on my device then. Should I remove my check in this MR ?

**Comment by Xyene** [comment]
Yeah, it'll prevent a merge conflict.

**Comment by stacyharper** [comment]
Okay I dont have any crashes with both. But it will cause build conflicts cause I removed the time arg from the default seatop begin. 

I can push this MR with the other commit. We could then merge them in the correct order. 

Should I ?

**Comment by Xyene** [comment]
Since https://github.com/swaywm/sway/pull/6418 has been merged, could you rebase this PR?

**Comment by stacyharper** [comment]
Okay I just rebased, resolved conflict as better as I tried and pushed it back.

I now got crashed. I guess the `container_raise_floating(con)` from `seatop_begin_down` still is in cause. I tried to add back my check condition and I dit crashed as soon as before. But I still had a crash later on my tries.

What should I know about https://github.com/swaywm/sway/pull/6418 ?

File: sway/input/seatop_down.c
Comment (stacyharper) [inline sway/input/seatop_down.c:201]: Here, should I try to use seatop_begin_down or the new seatop_begin_down_on_surface ?

Comment (stacyharper) [reply sway/input/seatop_down.c]: Okay I used this seatop_begin_down_on_surface

**Comment by stacyharper** [comment]
It looks like I fixed my crashes using directly seatop_begin_down_on_surface and not seatop_begin_down. The `con` in my context can be a `NULL` pointer so relying on it looks dangerous.

**Comment by stacyharper** [comment]
Nope, it crash when playing too much with swaybar :(

**Comment by stacyharper** [comment]
This line make Sway to crash `wl_signal_add(&e->surface->events.destroy, &e->surface_destroy);` when touching swaybar then draggin above when there is no window.

**Comment by stacyharper** [comment]
Okay I think I fixed my crashes:

- I had to toggle off `simulating_pointer_from_touch` before starting my seatop in cursor.c `handle_touch_down`. The reason is that it is possible to have a finger or a touch on an empty workspace before a layer shell pop out. Which stuck the cursor in emulation mode. Then the new touch_down touches will behave wrongly and cause crashes.
- I fixed my surface determination in the default seatop `handle_touch_down`. I used the cursor x and y which is wrong and cause a NULL pointer as surface. Attaching destroy event to this pointer was causing a crash

**Comment by stacyharper** [comment]
Hello there ! Someone got an idea of when this could be merged ? :D It prevent us in sxmo to chaim officialy the support of sway

File: sway/input/seatop_down.c
Comment (emersion) [inline sway/input/seatop_down.c:59]: Style: unnecessary `;`

File: sway/input/seatop_down.c
Comment (emersion) [inline sway/input/seatop_down.c:38]: Style: unnecessary `;`

File: sway/input/seatop_down.c
Comment (emersion) [inline sway/input/seatop_down.c:36]: What if the point wasn't found in the list? I'd prefer adding ` bool found` variable with `sway_assert(found)` if it's supposed to always be there.

File: sway/input/seatop_down.c
Comment (emersion) [inline sway/input/seatop_down.c:56]: This removes the point event but doesn't free it.

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c]: fixed

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c]: fixed

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c:38]: okay I added this check

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c:59]: I added a free(point_event), hoping its enough

**Comment by stacyharper** [comment]
I pushed a rebase I had in my tree. I use this tree for some month now without crashes

File: sway/input/cursor.c
Comment (emersion) [inline sway/input/cursor.c:528]: Why is this outside of the `seat->touch_id == event->touch_id` condition? It sounds like this potentially mixes up some touch points.

File: sway/input/seatop_down.c
Comment (emersion) [reply sway/input/seatop_down.c:59]: Yes, should be enough.

File: sway/input/seatop_down.c
Comment (emersion) [inline sway/input/seatop_down.c:69]: Nit: should we call this before `seatop_begin_default`?

**Comment by emersion** [comment]
Apart from these comments, this looks good to me (both code and testing).

File: sway/input/cursor.c
Comment (stacyharper) [reply sway/input/cursor.c]: We use the seat->touch_x/y as storage to the default seatop seatop_touch_motion implementation. As the seat->touch_id is the last touch down touch id, moving this inside the condition make both touches to use the last touch point positions.

But that feels more like a comfort choice, so I decided to revert it. So now seat->touch_x/y will only represent the last touch positions (as before). In the seatop_down, I then recompute lx and ly using the event x and y.

I let this change in a dedicated commit to make it easier for review. I'll fixup if everything looks okay to you.

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c:75]: To be honnest, I have no idea o_O

**Comment by stacyharper** [comment]
> Apart from these comments, this looks good to me (both code and testing).

One things remains to me after some of my last tries:

I added a little bit of logic to prevent the seatop_down to handle touches initiated from other surface than the initial one. But it looks like it doesnt works as much as I expect it. If I touch two time in another surface, it looks like it reset the seatop :S I'm not sure of what I am missing atm

**Comment by stacyharper** [comment]
> If I touch two time in another surface, it looks like it reset the seatop :S I'm not sure of what I am missing atm

to be more precise, my patch works for surfaces that handle touch events. By example, with two wev windows, I cannot interact with the second one when I have already a seatop_down on the first one. But in the case of web + foot, touching foot seems to stop the seatop_down. Maybe it is related to mouse simulation ? cause foot does not handle touch events yet

**Comment by stacyharper** [comment]
sway/input/cursor.c:481 handle_touch_down

```c
→   if (surface && wlr_surface_accepts_touch(wlr_seat, surface)) { 
→   →   if (seat_is_input_allowed(seat, surface)) { 
→   →   →   cursor->simulating_pointer_from_touch = false; 
→   →   →   seatop_touch_down(seat, event); 
→   →   } 
→   } else if (!cursor->simulating_pointer_from_touch && 
→   →   →   (!surface || seat_is_input_allowed(seat, surface))) { 
→   →   // Fallback to cursor simulation. 
→   →   // The pointer_touch_id state is needed, so drags are not aborted when over 
→   →   // a surface supporting touch and multi touch events don't interfere. 
→   →   cursor->simulating_pointer_from_touch = true; 
→   →   cursor->pointer_touch_id = seat->touch_id; 
→   →   double dx, dy; 
→   →   dx = lx - cursor->cursor->x; 
→   →   dy = ly - cursor->cursor->y; 
→   →   pointer_motion(cursor, event->time_msec, event->device, dx, dy, dx, dy); 
→   →   dispatch_cursor_button(cursor, event->device, event->time_msec, 
→   →   →   →   BTN_LEFT, WLR_BUTTON_PRESSED); 
→   } 
```

We should find a way to prevent the `else if` when the surface does not accept touches if a seatop already exists (if it is not the seatop_default ?)

**Comment by stacyharper** [comment]
> We should find a way to prevent the else if when the surface does not accept touches if a seatop already exists (if it is not the seatop_default ?)

Okay, in the last commit I moved the block inside the seatop_default. I also added the necessary check in the seatop_down (if seat can input or if the surface accept them). So now, while a seatop_down is running, we can only add new touches from the same initial surface until all touches are released.

**Comment by stacyharper** [comment]
@emersion does the last patches seems still good to you ? (I'm sorry for pinging but I feels like the MR tends to stall after the moment I send multiple comments to myself as I did)

File: sway/input/cursor.c
Comment (emersion) [inline sway/input/cursor.c:472]: The result of this functions seems unused.

**Comment by emersion** [comment]
Not sure I'll have time for a full review soon, I'll defer to @Xyene for now.

File: sway/input/cursor.c
Comment (stacyharper) [reply sway/input/cursor.c]: Yes you right ! I removed unessessary surface, sx and sy

**Comment by stacyharper** [comment]
This is a kindly up :+1: As it works finely and the code looks fine, could it be finally merged ? :D

**Comment by emersion** [comment]
We need someone to do a review first, as the code significantly changed since the last one.

**Comment by stacyharper** [comment]
I rebased this, fixed minor conflicts and adapted the event usage. It compile and still looks fine. I'll run it for sometime to see if there is no issue.

**Comment by stacyharper** [comment]
I bumped sxmo-sway to 1.8. This rebased patch continues to works nicely. I would love to drop sxmo-sway for the next release. Could someone plan to check this ?

**Comment by emersion** [comment]
Had a quick look, acked-by me

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c]: done

File: sway/input/seatop_down.c
Comment (stacyharper) [reply sway/input/seatop_down.c]: done

**Comment by stacyharper** [comment]
> Also `wlr_cursor_absolute_to_layout_coords` gets called twice in down/motion and their caller, this should be calculated once.

Add added lx and ly as arguments on touch_down and touch_motion. Also had to add them on the begin operation.

File: sway/input/seatop_down.c
Comment (bl4ckb0ne) [inline sway/input/seatop_down.c:220]: You copy pasted the log from the other `sway_assert`

Comment (stacyharper) [reply sway/input/seatop_down.c]: right, I didn't noticed the var name was used. Fixed

Comment (stacyharper) [reply sway/input/seatop_down.c]: I don't even know why this patch touch those lines anyway. I guess multiple rebase caused this. I should probably restore this simple return.

**Review by bl4ckb0ne — APPROVED** [review]

**Pull request merged**
